express/parsers/molecule.py

```diff
@@ -0,0 +1,109 @@
+from io import StringIO
+import os
+from ase.io import read, write
+import ase
+from rdkit import Chem
+import logging
+
+class MoleculeParser():
+    """
+    Molecule parser class.
+
+    Args:
+        args (list): args passed to the parser.
+        kwargs (dict): kwargs passed to the parser.
+            structure_string (str): structure string.
+            structure_format (str): structure format, poscar or espresso-in.
+    """
+
+    def __init__(self, structure_string):
+        self.structure_string = structure_string
+        self.is_using_inchi = self.is_pybel_import_successful()
+
+
+    def is_pybel_import_successful(self):
+        """
+        pybel is a submodule of openbabel used to generate the molecule string that
+        can then be converted into an InChI string by rdkit.
+
+        Pybel can be imported when the butler-venv is running.
+        Without the butler-venv, openbabel is not installed, thereby causing
+        the pybel import to fail.
+
+        When the butler-venv is not installed, inchi generation will be
+        set to 'None' and essentially skipped for the purpose of testing.
+
+        Without the 'try' statements present below ExPrESS will fail due to import errors.
+        If openbabel is installed properly pybel will be imported.
+        Otherwise, pybel import will fail and then get_inchi & get_inchi_key will
+        be set to 'None'
+
+        Returns:
+            Boolean
+        """
+        try:
+            import pybel
+            is_using_inchi = True
+        except ImportError:
+            is_using_inchi = False
+            logging.error("Pybel failed to import. InChI & InChI Key cannot be created.")
+        return is_using_inchi
+
+    def create_pybel_smile_from_poscar(self):
+        """
+        Function using ase to convert the POSCAR formatted string of a structure
+        into an XYZ formatted text file for that structure.
+
+        Then pybel converts the XYZ formatted text file into a SMILES format.
+
+        Returns:
+            Str: structure in SMILES format.
+        """
+        import pybel
+        xyz_file = "geom.xyz"
+        with open(xyz_file, "w") as file:
+            os.chmod(xyz_file, 0o777)
+            file_string = StringIO(self.structure_string)
+            ase_poscar = ase.io.read(file_string, format="vasp")
+            ase_xyz_file = ase.io.write(xyz_file, ase_poscar, format='xyz')
+            pybel_smile = list(pybel.readfile('xyz', 'geom.xyz'))[0]
```

Comment:
I really like the StringIO idea here - didn't know Python had these.

Reading/writing to the disk like this is really slow - especially if we're just throwing away the POSCAR/XYZ files right after. It might be worth changing that `ase.io.write` operation on line 68 to go into a StringIO object instead, and then using Pybel's readstring method. StringIO objects give access to their contents via the `getvalue()` method - https://open-babel.readthedocs.io/en/latest/UseTheLibrary/Python_PybelAPI.html#pybel.readstring

Not a big deal if this doesn't work, but would be a better than throwaway disk writes.

Reply:
I will give that a try. I'm not sure if it will work. If i can make it work I will though.

Reply:
The issue is that ase.io.write is written to work with files, not strings. 

ase.io.write does not return a string itself, but writes content to the file you pass it based on the input string you pass, as well as the format. There is not a good way to get from `POSCAR` to `XYZ` without going through an actual file, rather than a string. 

So for the time being i think that sticking with this disk write method is the best option. I can look into ways to fix this in the future.

Reply:
It'll take in any filepointer, so we can write it to a StringIO object, then just extract the string from it -

```python
In [1]: import ase.io, ase.cluster

In [2]: from io import StringIO

In [3]: outp = StringIO()

In [4]: cluster = ase.cluster.Icosahedron("Cu", 2)

In [5]: ase.io.write(outp, cluster)

In [6]: outp.getvalue()
Out[6]: '{"1": {\n "cell": {"array": {"__ndarray__": [[3, 3], "float64", [4.34283689515448, 0.0, 0.0, 0.0, 4.34283689515448, 0.0, 0.0, 0.0, 4.34283689515448]]}, "pbc": {"__ndarray__": [[3], "bool", [false, false, false]]}, "__ase_objtype__": "cell"},\n "ctime": 21.618267063645625,\n "mtime": 21.618267063645625,\n "numbers": {"__ndarray__": [[13], "int64", [29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29]]},\n "pbc": {"__ndarray__": [[3], "bool", [false, false, false]]},\n "positions": {"__ndarray__": [[13, 3], "float64", [2.17141844757724, 2.17141844757724, 2.17141844757724, 4.34283689515448, 2.17141844757724, 0.8294080431759738, 4.34283689515448, 2.17141844757724, 3.5134288519785057, 0.0, 2.17141844757724, 0.8294080431759738, 0.0, 2.17141844757724, 3.5134288519785057, 0.8294080431759738, 4.34283689515448, 2.17141844757724, 3.5134288519785057, 4.34283689515448, 2.17141844757724, 0.8294080431759738, 0.0, 2.17141844757724, 3.5134288519785057, 0.0, 2.17141844757724, 2.17141844757724, 0.8294080431759738, 4.34283689515448, 2.17141844757724, 3.5134288519785057, 4.34283689515448, 2.17141844757724, 0.8294080431759738, 0.0, 2.17141844757724, 3.5134288519785057, 0.0]]},\n "tags": {"__ndarray__": [[13], "int64", [1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]]},\n "unique_id": "895e1a2a8be97aa6846613f068d7b261",\n "user": "mat3ra"},\n"ids": [1],\n"nextid": 2}\n'
```

Just an idea, might be good as a future improvement if this is already ready to go.

Reply:
@adewyer - does @AcylSilane's suggestion above work to get rid of storing files?? If yes, please implement.